### ISSUE TITLE
No HTTP headers emitted by bin/cgi-bin/nph-genericinterface.pl

### ISSUE THREAD

**Issue by bschmalhofer** [RotherOSS/otobo#1104]
There are still SOAP related failures when the wrapper scripts in cgi-bin are used. An example for such an failing script is _scripts/test/GenericInterface/Operation/Ticket/TicketGet.t_ . The error message is:

> SOAP::Transport::HTTP::Client::send_receive: 500 Status line too long (limit is 8192)

This error message is emitted from `NET::HTTP::Methods::my_readline()` when trying to read in the HTTP header lines of the SOAP response. 

Let's first take a look whether the webservice is generating the expected result. For that, the test script need to  be adapted:
- Remove `TestMode => 1`from the webservice config in the TicketGet.t
- Remove the   `WebserviceDelete()` calls in the test script
- Limit the test script to the cases "SessionCreate" and "Test Ticket 4 With All Articles and Attachments". The latter case is the first test that actually fails.
- Copy the changed script into the top level otobo dir, for convenience

Now the Debugger in the web service Admin page can be used for inspecting the web service under test. The response generated on the server side looks fine. Hint: the XML response is longer than 8191 chars, HTTP headers are not displayed.  But parsing the response on the client side fails with the known error.

So, let's dig in deeper. Tracing can be enabled when `SOAP::Lite` is imported into `Kernel::GenericInterface::Transport::HTTP::SOAP`: `use SOAP::Lite +trace => 'all';`
Now the output from `prove -v -I . -I Kernel/cpan-lib/ TicketGet.t 2>&1 | tee prove.out` contains debugging information, including the HTTP headers that were sent and received. For comparison the output can be generated for the two relevant cases:

1. _scripts/apache2-httpd.include.conf_ linked into _/etc/apache2/sites-enabled_ (working)
2. _scripts/apache2-httpd.include.conf_  linked into _/etc/apache2/sites-enabled_ (with failures)

The relevant difference are in the response from the web service. Here is the response for SessionCreate:

re. 1.
```
Connection: close
Date: Sat, 26 Jun 2021 11:23:17 GMT
Server: Apache/2.4.41 (Ubuntu)
Vary: Accept-Encoding
Content-Length: 513
Content-Type: text/xml; charset=UTF-8
Client-Date: Sat, 26 Jun 2021 11:23:18 GMT
Client-Peer: 192.168.178.29:80
Client-Response-Num: 1

<?xml version="1.0" encoding="UTF-8"?><soap:Envelope soap:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"><soap:Body><SessionCreateResponse xmlns="http://otobo.org/SoapTestInterface/"><SessionID>fMsWLN1LmBh34AFVq3TQ5f3SH9Tn9V3O</SessionID></SessionCreateResponse></soap:Body></soap:Envelope>

```

re. 2.

> Client-Date: Sat, 26 Jun 2021 11:19:16 GMT
> Client-Peer: 192.168.178.29:80
> Client-Response-Num: 1
> 
> <?xml version="1.0" encoding="UTF-8"?><soap:Envelope soap:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"><soap:Body><SessionCreateResponse xmlns="http://otobo.org/SoapTestInterface/"><SessionID>S8TdJZSAWup9xoocFzcbtoiGhIi5JKQq</SessionID></SessionCreateResponse></soap:Body></soap:Envelope>
> 

The Client-* header fields are generated internally in the client library. So it looks like no HTTP headers are emitted at all in the case 2 . SOAP::Lite then assumes that the XML-Response contains headers and sometimes ran into 'Line to long'. So this looks like it is related to #1100.

_Originally posted by @bschmalhofer in https://github.com/RotherOSS/otobo/issues/1054#issuecomment-868998460_

**Comment by bschmalhofer** [comment]
Looked again on the server side using Plack::Middleware::DebugLogging. See  also https://github.com/mattp-/Plack-Middleware-DebugLogging/issues/5. Saw nothing significant there. Still no idea why no headers are seen in Kernel/GenericInterface::Transport::HTTP::SOAP. 
The next course of action: Run the webserver with the branch issue-#1054-bin_cgi_bin and the client with rel-10_0.

**Comment by bschmalhofer** [comment]
Finally starting to understand the different effects. The most relevant thing is the behaviour of ModPerl::Registry. In https://metacpan.org/release/SHAY/mod_perl-2.0.11/source/ModPerl-Registry/lib/ModPerl/RegistryCooker.pm we have the following code:
```

   # handle the non-parsed handlers ala mod_cgi (though mod_cgi does
    # some tricks removing the header_out and other filters, here we
    # just call assbackwards which has the same effect).
    my $base = File::Basename::basename($self->{FILENAME});
    my $nph = substr($base, 0, 4) eq 'nph-' ? '$_[0]->assbackwards(1);' : "";
    my $script_name = $self->get_script_name || $0;
 
    my $eval = join '',
                    'package ',
                    $self->{PACKAGE}, ";",
                    "sub handler {",
                    "local \$0 = '$script_name';",
                    $nph,
                    $shebang,
                    $line,
                    ${ $self->{CODE} },
                    "\n}"; # last line comment without newline?
```
This means that for CGI Skripts starting with **nph-** the Apache method `assbackwards()` is called. This methods directs Apache to pass thru any headers from the script directly to the client. This is the behavior expected for non parsed header scripts and is quite sensible. It is a bit strange that this behavior can't be turned off, but there are propably good reasons for that. It basically guarantees that any working scripts starting with **nph-** really is a non parsed header scripts. Another strangeness is of course that there is no good reason that nph-genericinterface.pl needs to be a non parsed header script.  AFAIK, there is no chunked output from the script. _Or is there chunked output?_

One consequence is that we need to reinstate the `PerlOptions -ParseHeaders` config for_ nph-genericinterface.pl_ in _scripts/apache2-httpd.include.conf_ .

Then we need to make sure that _nph-genericinterface.pl_ really prints out a valid HTTP header. See https://www.w3.org/Protocols/rfc2616/rfc2616-sec6.html. This might mean that Plack::Handler::CGI needs to be replaced by custom code. This is because the first line emitted by that module is:
`Status: 200 OK\r\n` instead of `HTTP1.1 200 OK\r\n`

**Comment by bschmalhofer** [comment]
Replacing the naive usage of Plack::Handler::CGI with adaped code derived from Plack::Handler::CGI did the trick. Test suite is completing now, but there are still unrelated failures. Changes will be merged with #1054. Closing this issue.

**Comment by bschmalhofer** [comment]
That `assbackward()` feature in `Modperl::Registry`is really not intuitive. It necessitates that nph-genericinterface.pl can't have the same simple implementation as e.g. index.pl. Therefore I propose to rename npn-genericinterface.pl to genericinterface.pl. For the 10.1 releases the old name can still be supported. But remove it for 11.0.
More importantly: nph- is so 1990s.
